src/Icicle/Avalanche/Statement/Simp/Eval.hs

```diff
@@ -0,0 +1,99 @@
+{-# LANGUAGE NoImplicitPrelude #-}
+{-# LANGUAGE PatternGuards     #-}
+module Icicle.Avalanche.Statement.Simp.Eval
+     ( simpEvalX
+     , simpEvalP
+     ) where
+
+import           Icicle.Common.Value
+import           Icicle.Common.Exp
+import           Icicle.Common.Type
+
+import           P
+
+
+simpEvalX   :: Ord n
+            => EvalPrim a n p
+            -> (p -> Type)
+            -> Exp a n p
+            -> Exp a n p
+simpEvalX ev ty = fixp (simpEvalX' ev ty)
+ where
+  fixp f x
+   | Just x' <- f x = fixp f x'
+   | otherwise      = x
+
+
+simpEvalX'  :: Ord n
+            => EvalPrim a n p
+            -> (p -> Type)
+            -> Exp a n p
+            -> Maybe (Exp a n p)
+simpEvalX' ev ty = go
+  where
+    go xx = case xx of
+      XApp a p q
+        | mp <- go p
+        , mq <- go q
+        , p' <- fromMaybe p mp
+        , q' <- fromMaybe q mq
+        , x' <- XApp a p' q'
+        , Just (prim, as) <- takePrimApps x'
+        , Just args       <- mapM (takeValue . just go) as
+        -> case simpEvalP ev ty a prim args of
+            Just x''
+             -> return x''
+            Nothing
+             | isJust mp || isJust mq
+             -> return x'
+             | otherwise
+             -> Nothing
+
+      XApp a p q
+        | Just p' <- go p, Just q' <- go q -> Just $ XApp a p' q'
+        | Just p' <- go p, Nothing <- go q -> Just $ XApp a p' q
+        | Nothing <- go p, Just q' <- go q -> Just $ XApp a p  q'
+        | otherwise                        -> Nothing
+
+      XLam a n t x1
+        -> XLam a n t <$> go x1
+
+      XLet a n p q
+        | Just p' <- go p, Just q' <- go q -> Just $ XLet a n p' q'
+        | Just p' <- go p, Nothing <- go q -> Just $ XLet a n p' q
+        | Nothing <- go p, Just q' <- go q -> Just $ XLet a n p  q'
+        | otherwise                        -> Nothing
+
+      XVar{}   -> Nothing
+      XPrim{}  -> Nothing
+      XValue{} -> Nothing
+
+    just f x = fromMaybe x (f x)
+
+
+-- | Primitive Simplifier
+--
+simpEvalP   :: Ord n
+            => EvalPrim a n p
+            -> (p -> Type)
+            -> a
+            -> p
+            -> [Value a n p]
+            -> Maybe (Exp a n p)
+simpEvalP ev ty a_fresh p vs
+ -- TODO: evaluation function should not return value for underapplied primitives
+ -- uncomment if you dare
+ | length (functionArguments $ ty p) == length vs
```

Comment:
ah.. so yeah, they are, but if you have
``pack value var``
it will look at "pack value var" and can't simp, because one of the arguments isn't a value. however, then it recurses down and sees "pack value", which is only applied to values…